### PR TITLE
refactor(mt): introduce BaseCachedTranslate class

### PR DESCRIPTION
**Pull request by miurahr**
## Pull request type

- Other (describe below)

refactor

## Which ticket is resolved?

Suggested in PR #470

## What does this PR change?

- Introduce `BaseCachedTranslate` class that has Cache layer.
- Change `BaseTranslate` not to handle cache

### CODE REVIEW

File: src/org/omegat/core/machinetranslators/BaseTranslate.java [hunk 1]
```diff
@@ -86,56 +73,10 @@ public BaseTranslate() {
...
-     * Creat cache object.
-     * <p>
-     * MT connectors can override cache size and invalidate policy.
-     * 
-     * @param name
-     *            name of cache which should be unique among MT connectors.
-     * @return Cache object
-     */
-    protected Cache<String, String> getCacheLayer(String name) {
-        return getCaffeineCache(name, 1_000, Duration.ONE_DAY);
```
Comment (miurahr) [inline src/org/omegat/core/machinetranslators/BaseTranslate.java:103]: Removed cache handling code from `BaseTranslate` class.

File: src/org/omegat/core/machinetranslators/BaseTranslate.java [hunk 2]
```diff
@@ -157,25 +98,77 @@ public void setGlossarySupplier(IMTGlossarySupplier glossarySupplier) {
...
     }
 
     @Override
     public String getCachedTranslation(Language sLang, Language tLang, String text) {
-        if (enabled) {
-            return getFromCache(sLang, tLang, text);
-        } else {
-            return null;
-        }
+        return null;
```
Comment (miurahr) [inline src/org/omegat/core/machinetranslators/BaseTranslate.java:109]: Always return null from `BaseTranslate`

File: src/org/omegat/core/machinetranslators/BaseTranslate.java [hunk 3]
```diff
@@ -157,25 +98,77 @@ public void setGlossarySupplier(IMTGlossarySupplier glossarySupplier) {
...
+     *            Source langauge.
+     * @param tLang
+     *            Target language.
+     * @param text
+     *            source text.
+     * @return translated text if exists in cache, otherwise null.
+     */
+    protected String getFromCache(Language sLang, Language tLang, String text) {
+        return null;
+    }
```
Comment (miurahr) [inline src/org/omegat/core/machinetranslators/BaseTranslate.java:135]: Keep method for backward binary compatibility.

File: src/org/omegat/core/machinetranslators/BaseTranslate.java [hunk 4]
```diff
@@ -285,6 +260,20 @@ protected boolean isCredentialStoredTemporarily(String id) {
...
+     *            original source text.
+     * @return truncated text.
+     */
+    protected String getTruncateText(String text) {
+        if (maxTextLength <= 0) {
+            return text;
+        }
+        return text.length() > maxTextLength ? text.substring(0, maxTextLength - 3) + "..." : text;
+    }
+
```
Comment (miurahr) [inline src/org/omegat/core/machinetranslators/BaseTranslate.java:276]: For future API stability.  see #473

**Review by miurahr — COMMENTED** [review]
These are the changes other than spotless coding style change.

File: src/org/omegat/core/machinetranslators/BaseTranslate.java (hunk 4 above)
Comment (miurahr) [reply src/org/omegat/core/machinetranslators/BaseTranslate.java:276]: It seems to be able to be private. I'd like to remove it.

**Review by t-cordonnier — CHANGES_REQUESTED** [review]
As said in another comment, I would recommend that in BaseCachedTranslate, methods `getTranslation` and `getCachedTranslation` become final, and all cache-related methods become private. This prevents the case where a user overrides getCachedTranslation in a way where the variable `cache` is still here but not used.
Users which don't agree with your cache policy can simply propose an alternative to BaseCachedTranslation class.

Except of that I approve this code.

**Comment by miurahr** [comment]
> all cache-related methods become private. 

@t-cordonnier  You cannot change `BaseCachedTranslate#getFromCache` and `BaseCachedTranslate#setToCache` method to be private without changing `BaseTranslate#putFromCache` and `BaseTranslate#putToCache`  that breaks binary compatibility against 3rd-party MT plugins that support OmegaT  5.7.1 and before.

Could you suggest a better way?

**Comment by miurahr** [comment]
@t-cordonnier  does #473 make you satisfied?

**Comment by t-cordonnier** [comment]
> @t-cordonnier You cannot change `BaseCachedTranslate#getFromCache` and `BaseCachedTranslate#setToCache` method to be private without changing `BaseTranslate#putFromCache` and `BaseTranslate#putToCache` that breaks binary compatibility against 3rd-party MT plugins that support OmegaT 5.7.1 and before.
> 
> Could you suggest a better way?

You can depreciate the method putToCache in BaseTranslate and use a new name in BaseCachedTranslate
Or even depreciate class BaseTranslate and start a new hierarchy

> @t-cordonnier does https://github.com/omegat-org/omegat/pull/473 make you satisfied?

Sorry when I reviewed it I thought it was identical except for adding maxTextLength handling
I will have a new look and tell you later

**Comment by miurahr** [comment]
A change #473 will be future figure and I think we reach a result;

- make putToCache and getFromCache deprecated  (in 6.1)
- make getCachedTranslate and getTranslate final (in 6.1)
- add getMaxTextLength()
- introduce BaseCachedTranslate class

I've changed last two here,  and plan to first two.

File: src/org/omegat/core/machinetranslators/BaseCachedTranslate.java [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public final String getCachedTranslation(Language sLang, Language tLang, String text) {
```
Comment (miurahr) [inline src/org/omegat/core/machinetranslators/BaseCachedTranslate.java:128]: Now the method is final. Adjusted from review comment.

**Review by t-cordonnier — APPROVED** [review]
Looks ok for me now

**Pull request merged**